src/nvim/buffer_defs.h

```diff
@@ -1158,6 +1158,12 @@ struct window_S {
    * In a non-location list window, w_llist_ref is NULL.
    */
   qf_info_T   *w_llist_ref;
+
+  /*
+   * Are we following the terminal? (Only valid if w_buffer is a terminal
+   * buffer.)
+   */
+  bool following_terminal;
```

Comment:
Indeed, we need to decide for each window whether it is following the terminal at that moment in time... A variable that's tied to the buffer is too broad.
If there are any suggestions for window-local variables that could be reused, I'll gladly switch. I had a quick look at what's already in the `struct window_S`, and could find nothing that's obviously suitable. OTOH my knowledge of those data structures is rather limited, so I could have missed something...